### ISSUE TITLE
macro fragment names are no longer idents

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#42019]
This is a stable-to-beta regression, but it's such a weird thing to do that I don't know if it counts.

Consider the following code ([playpen](https://is.gd/RmWrlU)):

```rust
macro_rules! gen {
    ($name:ident ($($dol:tt $var:ident)*) $($body:tt)*) => {
        macro_rules! $name {
            ($($dol $var:ident)*) => {
                $($body)*
            }
        }
    }
}
```

The `gen` macro parses a macro argument spec by matching a dollar sign with `$dol` and the fragment name with `$var`. Then (in this toy example) it just regurgitates that adding `:ident` as the type of each fragment.

This works in 1.17, but fails when invoked in 1.18 and nightly with the message "expected identifier, found [the ident passed in for `$var`]", pointing at `$var` on line 4. All three channels work if `$var:ident` is changed to `$var:tt` on line 2.

**Comment by durka** [comment]
cc @jseyfried

**Comment by jseyfried** [comment]
I believe this is a duplicate of #41803; it should be fixed once #42005 lands.

**Comment by nagisa** [comment]
Now that #42005 has landed, this should be re-reproduced.

**Comment by durka** [comment]
It landed, but why is there no nightly?

**Comment by durka** [comment]
This was **not fixed** by #42005. It is still broken with rustc 1.19.0-nightly (0ed1ec9f9 2017-05-18).

**Comment by brson** [comment]
Still broken on beta?

**Comment by durka** [comment]
Yes, the code in the playpen still fails to compile on beta and nightly.

**Comment by nikomatsakis** [comment]
cc @Mark-Simulacrum -- it'd be great to know just what caused this problem. @jseyfried, are you going to have time to investigate?

**Comment by Mark-Simulacrum** [comment]
If I have a chance I'll run bisect-rust, also cc @est31 -- they have it working locally as well.

**Comment by Mark-Simulacrum** [comment]
Ran bisect: fe151194e9a919ab23c2bb356767a5cbfa3811e7 - #40597.

**Comment by jseyfried** [comment]
Yeah, I'll investigate/fix this weekend.

**Comment by nikomatsakis** [comment]
@jseyfried any progress?

**Comment by brson** [comment]
This is a P-high bug and it is inactive cc @rust-lang/compiler

**Comment by jseyfried** [comment]
Fixed in #43224.
